### ISSUE TITLE
Building with CGAL from source

### ISSUE THREAD

**Issue by alecjacobson** [libigl/libigl#1434]
CGAL depends on gmp, mpfr, and boost

libigl's cmake routine will download cgal source, but as far as I can tell this will not contain gmp, mpfr, and boost  and will assume they're installed at the system level (e.g., with `apt-get` on linux or `brew` on mac os x).

On a machine without package manager / root access, it is still possible to install of this from source, albeit quite painful and storage intensive (3.3GB).

Here's what I did:

```
git clone http://github.com/libigl/libigl
cd libigl/
mkdir external
cd external/
wget https://gmplib.org/download/gmp/gmp-6.2.0.tar.xz
tar xvf gmp-6.2.0.tar.xz
wget https://www.mpfr.org/mpfr-current/mpfr-4.0.2.tar.xz
tar xvf mpfr-4.0.2.tar.xz mpfr-4.0.2/
wget https://dl.bintray.com/boostorg/release/1.66.0/source/boost_1_66_0.tar.gz
tar xvzf boost_1_66_0.tar.gz
cd gmp-6.2.0/
mkdir build
./configure --prefix build
make -j16
make install
cd ../mpfr-4.0.2/
mkdir build
./configure --prefix=`pwd`/build --with-gmp-include=../gmp-6.2.0/build/include --with-gmp-lib=../gmp-6.2.0/build/lib
make -j16
make install
cd ../boost_1_66_0
./bootstrap.sh
./b2 -j16
./b2 install --prefix=build
export BOOST_ROOT=`pwd`/build
rm $BOOST_ROOT/lib/libboost_*so*
rm $BOOST_ROOT/lib/libboost_*so
cd ../../
mkdir build
cd build
cmake ../ -DCMAKE_BUILD_TYPE=Release -DGMP_INCLUDE_DIR=`pwd`/../external/gmp-6.2.0 -DGMP_LIBRARIES=`pwd`/../external/gmp-6.2.0/build/lib/libgmp.so -DMPFR_INCLUDE_DIR=`pwd`/../external/mpfr-4.0.2/build/include -DMPFR_LIBRARIES=`pwd`/../external/mpfr-4.0.2/build/lib/libmpfr.a
make -j16 libigl_tests
./libigl_tests

```

It would be amazing if this could be automated like the other dependencies during the libigl cmake build.

**Comment by danielepanozzo** [comment]
I think @jdumas already figured this out on windows. Not sure about macosx/linux.

**Comment by jdumas** [comment]
Yes, so our CMake routine downloads gmp and mpfr on Windows, this is happening [here](https://github.com/libigl/libigl/blob/4ce917d424efd941174c6ff46adf15790c2ad12a/cmake/libigl.cmake#L129). The thing is that it works because those are precompiled dlls, and we set up the `ENV` variables `GMP_DIR` and `MPFR_DIR` which are looked up by CGAL [here](https://github.com/CGAL/cgal/blob/f7c3c8212b56c0d6dae63787efc99093f4383415/Installation/cmake/modules/FindGMP.cmake#L29) and [here](https://github.com/CGAL/cgal/blob/f7c3c8212b56c0d6dae63787efc99093f4383415/Installation/cmake/modules/FindMPFR.cmake#L28).

For boost, I made this [conda environment file](https://github.com/libigl/libigl/blob/dev/cmake/libigl-cgal.yml) for Windows users, but it should also work to install boost (without having to compile it) on macOS. We might have to set the `ENV` variable `BOOST_ROOT` to point to `CONDA_PREFIX` so that CMake's FindBoost.cmake is able to find the version of boost installed in conda.

I believe the conda environment could be used to install a precompiled version of gmp and mpfr on macOS and Linux as well (seeing as the packages are available on `conda-forge`). We might need to adjust our CMake to set the `ENV` variables for `GMP_DIR` and `MPFR_DIR` accordingly (e.g. if those are undefined, and the `CONDA_PREFIX` environment variable exists).
